File: Currency.cs
// Currency.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

class CurrencyConverter
{
    private static readonly HttpClient client = new HttpClient();
    private static readonly string ApiUrl = "https://api.exchangerate.host/latest?base=USD";
    private static readonly TimeSpan CacheTTL = TimeSpan.FromSeconds(60);
    private static Dictionary<string, double> _rates;
    private static DateTime _cacheTime;
    private static List<(double amount, string from, string to, double result, string time)> _history = new();

    static async Task<Dictionary<string, double>> GetRates()
    {
        var now = DateTime.UtcNow;
        if (_rates != null && (now - _cacheTime) < CacheTTL)
            return _rates;

        var response = await client.GetAsync(ApiUrl);
        response.EnsureSuccessStatusCode();
        var json = await response.Content.ReadAsStringAsync();
        var data = JsonSerializer.Deserialize<RatesResponse>(json);
        if (!data.Success)
            throw new Exception("API error");
        _rates = data.Rates;
        _cacheTime = now;
        return _rates;
    }

    public static async Task<double> ConvertAsync(double amount, string fromCur, string toCur)
    {
        var rates = await GetRates();
        if (!rates.ContainsKey(fromCur))
            throw new Exception($"Currency '{fromCur}' not supported.");
        if (!rates.ContainsKey(toCur))
            throw new Exception($"Currency '{toCur}' not supported.");
        double usdAmount = fromCur == "USD" ? amount : amount / rates[fromCur];
        return toCur == "USD" ? usdAmount : usdAmount * rates[toCur];
    }

    static void AddHistory(double amount, string fromCur, string toCur, double result)
    {
        _history.Add((amount, fromCur, toCur, result, DateTime.Now.ToString("HH:mm:ss")));
        if (_history.Count > 10) _history.RemoveAt(0);
    }

    static void ShowHistory()
    {
        if (_history.Count == 0)
        {
            Console.WriteLine("No conversions yet.");
            return;
        }
        Console.WriteLine("\n--- History ---");
        foreach (var h in _history)
            Console.WriteLine($"{h.time}  {h.amount:F2} {h.from} = {h.result:F2} {h.to}");
    }

    static async Task Interactive()
    {
        Console.WriteLine("=== Currency Converter ===");
        while (true)
        {
            Console.WriteLine("\n1. Convert");
            Console.WriteLine("2. Show history");
            Console.WriteLine("3. Exit");
            Console.Write("Choose: ");
            string choice = Console.ReadLine()?.Trim() ?? "";
            switch (choice)
            {
                case "1":
                    Console.Write("Amount: ");
                    if (!double.TryParse(Console.ReadLine(), out double amount) || amount < 0)
                    {
                        Console.WriteLine("Invalid amount.");
                        continue;
                    }
                    Console.Write("From (USD): ");
                    string fromCur = Console.ReadLine()?.Trim().ToUpper() ?? "USD";
                    if (string.IsNullOrEmpty(fromCur)) fromCur = "USD";
                    Console.Write("To: ");
                    string toCur = Console.ReadLine()?.Trim().ToUpper() ?? "";
                    if (string.IsNullOrEmpty(toCur))
                    {
                        Console.WriteLine("Please enter a target currency.");
                        continue;
                    }
                    try
                    {
                        double result = await ConvertAsync(amount, fromCur, toCur);
                        Console.WriteLine($"{amount:F2} {fromCur} = {result:F2} {toCur}");
                        AddHistory(amount, fromCur, toCur, result);
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine($"Error: {e.Message}");
                    }
                    break;
                case "2":
                    ShowHistory();
                    break;
                case "3":
                    Console.WriteLine("Goodbye!");
                    return;
                default:
                    Console.WriteLine("Invalid choice.");
                    break;
            }
        }
    }

    static async Task Cli(string[] args)
    {
        if (args.Length != 3)
        {
            Console.WriteLine("Usage: Currency.exe <amount> <from_currency> <to_currency>");
            return;
        }
        if (!double.TryParse(args[0], out double amount) || amount < 0)
        {
            Console.WriteLine("Invalid amount.");
            return;
        }
        string fromCur = args[1].ToUpper();
        string toCur = args[2].ToUpper();
        try
        {
            double result = await ConvertAsync(amount, fromCur, toCur);
            Console.WriteLine($"{amount:F2} {fromCur} = {result:F2} {toCur}");
        }
        catch (Exception e)
        {
            Console.WriteLine($"Error: {e.Message}");
        }
    }

    static async Task Main(string[] args)
    {
        if (args.Length == 0)
            await Interactive();
        else
            await Cli(args);
    }

    class RatesResponse
    {
        public bool Success { get; set; }
        public Dictionary<string, double> Rates { get; set; }
    }
}

File: Currency.java
// Currency.java
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.TimeUnit;
import com.google.gson.*;

public class Currency {
    private static final String API_URL = "https://api.exchangerate.host/latest?base=USD";
    private static final long CACHE_TTL_MS = TimeUnit.SECONDS.toMillis(60);
    private static Map<String, Double> rates = null;
    private static long cacheTime = 0;
    private static List<HistoryEntry> history = new ArrayList<>();
    private static Scanner scanner = new Scanner(System.in);
    private static Gson gson = new Gson();

    static class RatesResponse {
        boolean success;
        Map<String, Double> rates;
    }

    static class HistoryEntry {
        double amount;
        String fromCur, toCur;
        double result;
        String time;
        HistoryEntry(double a, String f, String t, double r, String tm) {
            amount = a; fromCur = f; toCur = t; result = r; time = tm;
        }
    }

    static Map<String, Double> getRates() throws Exception {
        long now = Instant.now().toEpochMilli();
        if (rates != null && (now - cacheTime) < CACHE_TTL_MS)
            return rates;

        URL url = new URL(API_URL);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setConnectTimeout(5000);
        conn.setReadTimeout(5000);
        if (conn.getResponseCode() != 200) {
            throw new Exception("HTTP error: " + conn.getResponseCode());
        }
        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = br.readLine()) != null) sb.append(line);
            RatesResponse resp = gson.fromJson(sb.toString(), RatesResponse.class);
            if (!resp.success) throw new Exception("API error");
            rates = resp.rates;
            cacheTime = now;
            return rates;
        }
    }

    static double convert(double amount, String fromCur, String toCur) throws Exception {
        Map<String, Double> rates = getRates();
        if (!rates.containsKey(fromCur))
            throw new Exception("Currency '" + fromCur + "' not supported.");
        if (!rates.containsKey(toCur))
            throw new Exception("Currency '" + toCur + "' not supported.");
        double usdAmount = fromCur.equals("USD") ? amount : amount / rates.get(fromCur);
        return toCur.equals("USD") ? usdAmount : usdAmount * rates.get(toCur);
    }

    static void addHistory(double amount, String fromCur, String toCur, double result) {
        history.add(new HistoryEntry(amount, fromCur, toCur, result, 
            java.time.LocalTime.now().format(java.time.format.DateTimeFormatter.ofPattern("HH:mm:ss"))));
        if (history.size() > 10) history.remove(0);
    }

    static void showHistory() {
        if (history.isEmpty()) {
            System.out.println("No conversions yet.");
            return;
        }
        System.out.println("\n--- History ---");
        for (HistoryEntry h : history)
            System.out.printf("%s  %.2f %s = %.2f %s%n", h.time, h.amount, h.fromCur, h.result, h.toCur);
    }

    static void interactive() throws Exception {
        System.out.println("=== Currency Converter ===");
        while (true) {
            System.out.println("\n1. Convert");
            System.out.println("2. Show history");
            System.out.println("3. Exit");
            System.out.print("Choose: ");
            String choice = scanner.nextLine().trim();
            switch (choice) {
                case "1":
                    System.out.print("Amount: ");
                    String amountStr = scanner.nextLine().trim();
                    double amount;
                    try { amount = Double.parseDouble(amountStr); } catch (NumberFormatException e) {
                        System.out.println("Invalid amount.");
                        continue;
                    }
                    if (amount < 0) { System.out.println("Invalid amount."); continue; }
                    System.out.print("From (USD): ");
                    String fromCur = scanner.nextLine().trim().toUpperCase();
                    if (fromCur.isEmpty()) fromCur = "USD";
                    System.out.print("To: ");
                    String toCur = scanner.nextLine().trim().toUpperCase();
                    if (toCur.isEmpty()) { System.out.println("Please enter a target currency."); continue; }
                    try {
                        double result = convert(amount, fromCur, toCur);
                        System.out.printf("%.2f %s = %.2f %s%n", amount, fromCur, result, toCur);
                        addHistory(amount, fromCur, toCur, result);
                    } catch (Exception e) {
                        System.out.println("Error: " + e.getMessage());
                    }
                    break;
                case "2":
                    showHistory();
                    break;
                case "3":
                    System.out.println("Goodbye!");
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }

    static void cli(String[] args) throws Exception {
        if (args.length != 3) {
            System.out.println("Usage: java Currency <amount> <from_currency> <to_currency>");
            return;
        }
        double amount;
        try { amount = Double.parseDouble(args[0]); } catch (NumberFormatException e) {
            System.out.println("Invalid amount.");
            return;
        }
        if (amount < 0) { System.out.println("Invalid amount."); return; }
        String fromCur = args[1].toUpperCase();
        String toCur = args[2].toUpperCase();
        try {
            double result = convert(amount, fromCur, toCur);
            System.out.printf("%.2f %s = %.2f %s%n", amount, fromCur, result, toCur);
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length == 0)
            interactive();
        else
            cli(args);
    }
}
